### PR TITLE
Create boot.loader.systemd-boot.rebootForBitlocker option

### PR DESCRIPTION
**Pull request by thblt**
## Description of changes

This creates a `boot.loader.systemd-boot.rebootForBitlocker` option

Windows with BitLocker and TPM enabled doesn't support boot chaining. This option activates a special experimental mode in systemd-boot that tries to detect such systems and, if detected and selected by the user at the boot menu, set the BootNext EFI variable to it before resetting.

I suspect the maintainers of systemd may want to be notified:  @flokli @kloenk @mic92
## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thblt** [comment]
This is still waiting for feedback. @dasJ, as code owner, what do you think?

**Review by kloenk — APPROVED** [review]
Don't have a setup to test right now, but code looks reasonable

**Comment by shymega** [comment]
LGTM, but cautious.

I'd like to get this merged (and backported to 24.05) but with my experiments with adding `reboot-for-bitlocker yes` to my own NixOS configuration, I've noted it's not a supported option in `systemd-boot`.

I've rebased this PR against mainline and may open a new PR based on the rebased version of this PR. I'm just cautious that the current, version of `systemd` - which, as we know, bundles `systemd-boot` in the same package - in 24.05, does not include this configuration option.

Therefore, I believe we can merge, but we should keep it in `nixpkgs-unstable`, and aim for release in 24.11.

Thoughts?

**Review by shymega — APPROVED** [review]
LGTM., but we may have to restrict release to unstable, or 24.11 branch.

**Comment by shymega** [comment]
Actually, having said all that, `systemd` on NixOS 24.05 is on 255, and according to the man page for `loader.conf`, the Bitlocker option was introduced in version 251. So, perhaps we just need to enable a compile-time option. However, again, the question is - do we backport to 24.05?

And if so, will we put a strain on Hydra and other caches?

**Comment by flokli** [comment]
I don't quite understand the talks about compile-time options and strain on hydra.

The `reboot-for-bitlocker` is not gated by a compile-time flag in systemd, it's already available today (if you manually edit `loader.conf`. It's also documented in the `loader.conf` manpage, for that matter.

This PR simply exposes a NixOS module system option to make that string appear, it's not causing any rebuilds (which is why it initially targeted master). As the option stays dormant if not set, tthis could also be backported to 23.05.

I rebased this PR to match the changes that happened in these files in the meantime. PTAL.

**Review by nikstur — APPROVED** [review]
Looks fine to me. Rebuilding is not a concern here. 

Maybe we should eventually expose a settings option for `loader.conf` so we don't have re-implement interfaces for every feature.

**Comment by flokli** [comment]
Seems the automatic backport failed due to conflicts.

@shymega in case you want this backported, please open a PR and highlight me, happy to take a look.

**Comment by shymega** [comment]
Sorry, didn't get round to this. I have updated to latest 24.05, but I get the same error (despite updating `systemd-boot` with `--install-bootloader`) I had before with the `reboot-for-bitlocker yes` option set manually.

I'll open a bug report in a moment.

The output I see is:

```
$ sudo nixos-rebuild boot

building the system configuration... /boot/efi/NIXOS/loader/loader.conf:3: Unknown line 'reboot-for-bitlocker', ignoring.
```

**Pull request merged**
